src/shadowbox/server/shared_metrics.ts

```diff
@@ -82,18 +94,61 @@ export class PrometheusUsageMetrics implements UsageMetrics {
 
   constructor(private prometheusClient: PrometheusClient) {}
 
+  private async queryUsage(
+    timeSeriesSelector: string,
+    deltaSecs: number
+  ): Promise<PrometheusQueryResult> {
+    const query = `
+      sum(increase(${timeSeriesSelector}[${deltaSecs}s]))
+      by (${PROMETHEUS_COUNTRY_LABEL}, ${PROMETHEUS_ASN_LABEL})
+    `;
+    const queryResponse = await this.prometheusClient.query(query);
+    const result: PrometheusQueryResult = {};
+    for (const entry of queryResponse.result) {
+      const serializedKey = JSON.stringify(entry.metric, Object.keys(entry.metric).sort());
+      result[serializedKey] = {
+        metric: entry.metric,
+        value: Math.round(parseFloat(entry.value[1])),
+      };
+    }
+    return result;
+  }
+
   async getLocationUsage(): Promise<LocationUsage[]> {
     const timeDeltaSecs = Math.round((Date.now() - this.resetTimeMs) / 1000);
-    // We measure the traffic to and from the target, since that's what we are protecting.
-    const result = await this.prometheusClient.query(
-      `sum(increase(shadowsocks_data_bytes_per_location{dir=~"p>t|p<t"}[${timeDeltaSecs}s])) by (location, asn)`
-    );
-    const usage = [] as LocationUsage[];
-    for (const entry of result.result) {
-      const country = entry.metric['location'] || '';
-      const asn = entry.metric['asn'] ? Number(entry.metric['asn']) : undefined;
-      const inboundBytes = Math.round(parseFloat(entry.value[1]));
-      usage.push({country, inboundBytes, asn});
+    const [dataBytesResult, tunnelTimeResult] = await Promise.all([
+      // We measure the traffic to and from the target, since that's what we are protecting.
+      this.queryUsage('shadowsocks_data_bytes_per_location{dir=~"p>t|p<t"}', timeDeltaSecs),
+      this.queryUsage('shadowsocks_tunnel_time_seconds_per_location', timeDeltaSecs),
+    ]);
+
+    // We join the bytes and tunneltime metrics together by location (i.e. country and ASN).
+    const mergedResult: {
+      [metricKey: string]: {
+        metric: QueryResultMetric;
+        inboundBytes?: number;
+        tunnelTimeSec?: number;
+      };
+    } = {};
+    for (const [key, entry] of Object.entries(dataBytesResult)) {
+      mergedResult[key] = {...mergedResult[key], metric: entry.metric, inboundBytes: entry.value};
+    }
+    for (const [key, entry] of Object.entries(tunnelTimeResult)) {
+      mergedResult[key] = {...mergedResult[key], metric: entry.metric, tunnelTimeSec: entry.value};
+    }
+
+    const usage: LocationUsage[] = [];
+    for (const entry of Object.values(mergedResult)) {
+      const country = entry.metric[PROMETHEUS_COUNTRY_LABEL] || '';
+      const asn = entry.metric[PROMETHEUS_ASN_LABEL]
+        ? Number(entry.metric[PROMETHEUS_ASN_LABEL])
+        : undefined;
+      usage.push({
+        country,
+        asn,
+        inboundBytes: entry.inboundBytes || 0,
```

Comment:
Does it make a difference whether this is undefined or zero?

Will the backend fail if tunnelTimeSec is present?